File: Binary_Search/bin_Answer/q9_437-copy-books.py
# 	Q9_437. 书籍复印
# 给定 n 本书, 第 i 本书的页数为 pages[i]. 现在有 k 个人来复印这些书籍, 而每个人只能复印编号连续的一段的书, 比如一个人可以复印 pages[0], pages[1], pages[2], 但是不可以只复印 pages[0], pages[2], pages[3] 而不复印 pages[1].
# 所有人复印的速度是一样的, 复印一页需要花费一分钟, 并且所有人同时开始复印. 怎样分配这 k 个人的任务, 使得这 n 本书能够被尽快复印完?
# 返回完成复印任务最少需要的分钟数.
# 示例 1：
# 输入: pages = [3, 2, 4], k = 2
# 输出: 5
# 解释: 第一个人复印前两本书, 耗时 5 分钟. 第二个人复印第三本书, 耗时 4 分钟.
# 样例 2:
# 输入: pages = [3, 2, 4], k = 3
# 输出: 4
# 解释: 三个人各复印一本书.
# 提示：
# 	书籍页数总和小于等于2147483647。
# 挑战：
# 	时间复杂度 O(nk)
# 相关： 1219. 加热器
#
# 思路: 使用九章算法强化班中所讲的基于答案值域的二分法。
# 答案的范围在 max(pages)~cnt(pages) 之间，每次二分到一个时间 time_limit 的时候，用贪心法从左到右扫描一下 pages，看看需要多少个人来完成复印。
# 如果这个值 <= k，那么意味着大家花的时间可能可以再少一些，如果 > k 则意味着人数不够，需要降低工作量。
# O(n log(cnt) )是该问题时间复杂度上的最优解法.

class Solution:
    def copyBooks(self, pages, k):
        if not pages: return 0
        left, right = max(pages), sum(pages)
        while left + 1 < right:
            mid = left + right >> 1
            if self.check(pages, mid) <= k:  # 所需人数在当前mid限制下，还可以再少
                right = mid  # 说明复印时间能够再少，查找左区间（mid以左）
            else:
                left = mid  # 在mid限制下，人数不够，说明复印时间相应增加（mid以右）
        if self.check(pages, left) <= k:
            return left
        return right

    def check(self, pages, time_limit):
        cnt, time_cost = 1, 0
        for page in pages:
            if page + time_cost > time_limit:
                cnt += 1
                time_cost = 0
            time_cost += page
        return cnt

File: Binary_Search/bin_Answer/q9_437_copy_books.java
package Binary_Search.bin_Answer;

import java.util.Arrays;

public class q9_437_copy_books {
    private int[] pages;
    private int k, n_book;
    public int copyBooks(int[] pages, int k) {
        if (pages == null || pages.length == 0 || k <= 0) return 0;
        this.pages = pages; this.k = k; this.n_book = pages.length;

        int left = Arrays.stream(pages).max().getAsInt();
        int right = Arrays.stream(pages).sum();
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            // 说明单机时间mid还可以减小(左区间)，从而增加工人k
            if (check(mid) <= k) { // 法2：[L, mid] & [mid+1, R](确保mid与L在一个区间)
                right = mid; // ❤优先左区间(∵求至少)
            } else left = mid;
        }
        // ∵至少 ∴优先左区间
        if (check(left) <= k) return left;
        return right;

        // 法2
//        while (left < right) { // [L, mid] & [mid+1, R](确保mid与L在一个区间)
//            int mid = left + right >> 1; // left + (right - left) >> 1;
//            if (check(mid) <= k) right = mid;
//            else left = mid + 1;
//        }
//        if (check(left) <= k) return left;
//        return -1;
    }

    private int check(int tryTimeLimit){
        int tryK = 1, total_time = 0;
        for (int page: pages) {
            if (page + total_time > tryTimeLimit) {
                tryK++; // 工人数++
                total_time = 0; // 当前工人的工作时间，初始化为 0
            } // ❤置零后，↓ 还需加上当前page，算作当前工人的工作
            total_time += page;
        }
        return tryK;
    }

    private int check2(int tryTimeLimit){
        int tryCnt = 1, total_time = 0;
        for (int i = 0; i < n_book; i++) {
            if (total_time + pages[i] <= tryTimeLimit) {
                total_time += pages[i];
            } else {
                tryCnt++; // 工人数++
                total_time = 0 + pages[i];
                // ❤置零后，↑ 还需加上当前pages[i]，算作当前工人的工作
            }
        }
        return tryCnt; // 在tryTimeLimit下，至少所需工人数
    }


}
